engine/kv_engine_cleaner.cpp

```diff
@@ -10,24 +10,40 @@ namespace KVDK_NAMESPACE {
 
 constexpr uint64_t kForegroundUpdateSnapshotInterval = 1000;
 
-template <typename T>
-T* KVEngine::removeListOutDatedVersion(T* list, TimeStampType min_snapshot_ts) {
-  static_assert(
-      std::is_same<T, List>::value || std::is_same<T, HashList>::value,
-      "Invalid collection type, should be list or hashlist.");
-  T* old_list = list;
-  while (old_list && old_list->GetTimeStamp() > min_snapshot_ts) {
-    old_list = old_list->OldVersion();
-  }
-
-  // the snapshot should access the old record, so we need to purge and free the
-  // older version of the old record
-  if (old_list && old_list->OldVersion()) {
-    auto older_list = old_list->OldVersion();
-    old_list->RemoveOldVersion();
-    return older_list;
-  }
-  return nullptr;
+template <typename ListType>
+void KVEngine::removeOutdatedList(ListType* collection) {
+  auto cur_id = collection->ID();
+  auto cur_collection = collection;
+  while (cur_collection) {
+    auto old_collection = cur_collection->OldVersion();
+    if (old_collection) {
+      auto old_collection_id = old_collection->ID();
+      if (old_collection_id != cur_id) {
+        cur_collection->RemoveOldVersion();
```

Comment:
Potential memory leak. Expired lists are chained after the list on HashTable. This function template only set the pointers to the older version(expired or deleted) as nullptr, but not store the pointer somewhere for later deletion.

Reply:
Also elements expired together with the List is not purged from PMem.